staging/kos/pkg/controllers/connectionagent/connection_agent.go

```diff
@@ -0,0 +1,1180 @@
+	// one exists.
+	var ifcOwner *netv1a1.NetworkAttachment
+	if indexer != nil {
+		ifcOwnerObj, err := indexer.ByIndex(ifcOwnerDataIndexerName, ifc.index())
+		if err != nil {
+			return fmt.Errorf("ByIndex(%s, %s) failed: %s", ifcOwnerDataIndexerName, ifc.index(), err.Error())
+		}
+		if len(ifcOwnerObj) == 1 {
+			ifcOwner, _ = ifcOwnerObj[0].(*netv1a1.NetworkAttachment)
+		}
+	}
+
+	if ifcOwner != nil {
+		ifcOwnerNSN := parse.AttNSN(ifcOwner)
+		_, ownerAlreadyHasInterface := ca.getNetworkInterface(ifcOwnerNSN)
+
+		// Pre-existing Network interfaces are matched to network attachments
+		// on the basis of the VNI, the IP and the host of the attachment. But
+		// these fields can be seen changing even if the namespaced name is
+		// steady, for instance by deleting the network attachment and creating
+		// one with same namespaced name and a different VNI. This means that
+		// two network interfaces could be matched to the same namespaced name,
+		// where each match would correspond to two different versions of the
+		// same network attachment, or two different network attachments with
+		// the same namespaced name. In such cases, the interface to keep is the
+		// one matching the most recent version of the network attachment, while
+		// the other should be deleted. Notice that the order in which the
+		// matches took place does not say which match should be kept, because
+		// the two versions of the matched network attachment might come from
+		// different informers and there are no cross-informer ordering
+		// guarantees. There are ways to always take the optimal choice, but
+		// they make for really complex code and yield little advantage because
+		// attachments can change: if an attachment is matched and it changes 1
+		// sec later, the match was useless. For this reason, this code does not
+		// attempt to always take the optimal choice: in cases of collisions the
+		// interface that was matched first is taken and the other is deleted.
+		// Collisions should be a rare event anyway. Even if the wrong choice is
+		// taken during normal operation the connection agent will rectify the
+		// the mistake by realizing the network interface does not match the
+		// attachment and creating a correct interface after deleting the old one.
+		if !ownerAlreadyHasInterface {
+			ca.assignNetworkInterface(ifcOwnerNSN, ifc)
+			klog.V(3).Infof("Matched pre-existing network interface %s with attachment %s", ifc, ifcOwnerNSN)
+			if localIfc, ifcIsLocal := ifc.(*localNetworkInterface); ifcIsLocal {
+				ca.setExecReport(ifcOwnerNSN, localIfc.id, ifcOwner.Status.PostCreateExecReport)
+				localIfc.postDeleteExec = ifcOwner.Spec.PostDeleteExec
+			} else {
+				// The interface is remote and so is its owner. There's no
+				// guarantee that the add notification handler for the owner has
+				// executed yet, neither as for when it will execute. If that
+				// has not happened by the time a worker processes the deletion
```

Comment:
220befb fixes that but introduces an error as well, efa1cad fixes that error.

Reply:
I did a review of [client-go/tools/cache](https://godoc.org/k8s.io/client-go/tools/cache) internals, and from what I saw the statements

> A SharedIndexedInformer has the property that every change that reaches its cache will be delivered in notifications to (at least) clients that were registered before the informer started running. Closing the stop channel of an informer primarily stops ingestion of changes. There are secondary implementation details of how goroutines terminate, and the short answer is that they do not terminate before their work is done.

are false; hence their consequence:

> Thus we can be certain that there will be a notification about ifcOwnerNSN by the informer for the VNI at hand. So we do not need to enqueue a reference here.

is false too.

An informer does not guarantee that the notification handler for a change that hit the cache executes even if its stop channel is closed. Hence the enqueue (or one of the other countermeasures we discussed) is needed.

Follows a detailed explanation. Any time a SharedIndexInformer's cache is mutated, the change is delivered to the informer's shared processor (e.g. [this](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/cache/shared_informer.go#L467) is the call for creation of objects), which multiplexes it to all the listeners (which execute the notification handlers). A [listener](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/cache/shared_informer.go#L570-L572) has two channels (`nextCh` and `addCh`) and is associated to two goroutines which execute the listener's [pop](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/cache/shared_informer.go#L614) and [run](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/cache/shared_informer.go#L644) methods respectively. `run` [reads a notification from nextCh and executes the appropriate notification handler function](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/cache/shared_informer.go#L653-L660), it returns when `nextCh` is closed. `pop` mainly consists of a for loop whose body has a select statement that attempts to:
1) [Send a notification to `nextCh`](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/cache/shared_informer.go#L622).
2) [Read a notification from `addCh`](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/cache/shared_informer.go#L629).

notifications are sent by the informer' shared processor to each listener through the listener's `addCh` ([here](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/cache/shared_informer.go#L513-L515) and [here](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/cache/shared_informer.go#L610)), when a listener receives a notification from `addCh`, it sets up the newly received notification to be sent on `nextCh` in a future for loop iteration or writes the notification to a buffer of pending notifications (from which it will later be read) if there is already a notification which it's trying to send down `nextCh`, [here](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/cache/shared_informer.go#L633-L639)'s the code. When a SharedIndexInformer's stopCh is closed, a domino effect where multiple stop channels are closed and internal components (e.g. reflector, etc...) are stopped takes place, and eventually it gets to the point where listeners are stopped.  Listeners are stopped by closing their `addCh`([here](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/cache/shared_informer.go#L532-L534)). More precisely, this [causes the goroutine that runs `pop` to terminate](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/cache/shared_informer.go#L629-L632) and the [listener's `nextCh` to be closed](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/cache/shared_informer.go#L616), which in turns causes the goroutine that runs `run` to terminate. Assume that an object _O_ has been created, and the listener's `pop` goroutine is waiting trying to send _O_'s creation notification to `nextCh`, but that is stalling because the execution of a handler for a previous notification in `run` is slow. At this point in time, _O_ is already in the SharedIndexInformer's cache but its creation notification handler has not executed yet. If the SharedIndexInformer's `stopCh` is closed and as a consequence `addCh` is closed before the send on _nextCh_ is performed, the listener is stopped and _O_'s creation notification will never execute.

Reply:
On the basis of my previous comment, fed0600 re-adds the enqueue of namespaced name X after matching a pre-existing remote network interface to an attachment with namespaced name X.